### PR TITLE
Use external provided scratchpad

### PR DESCRIPTION
**Pull request by junrushao**

### CODE REVIEW

**Review by yzh119 — COMMENTED** [review]
Just want to confirm, `TVM_DLL_EXPORT_TYPED_FUNC` is no longer required, is that correct?

File: src/tvm_wrapper.cu [hunk 1]
```diff
@@ -97,10 +75,11 @@ cudaError_t _SinglePrefillWithKVCacheNoLSE(DTypeIn* q, DTypeIn* k, DTypeIn* v, D
   return cudaSuccess;
 }
 
-int _FlashInferSinglePrefillWithKVCache(DLTensor* q, DLTensor* k, DLTensor* v, bool causal,
-                                        int64_t qkv_layout, int64_t rotary_mode,
+int _FlashInferSinglePrefillWithKVCache(DLTensor* q, DLTensor* k, DLTensor* v, DLTensor* tmp,
+                                        bool causal, int64_t qkv_layout, int64_t rotary_mode,
                                         bool allow_fp16_qk_reduction, double rope_scale,
                                         double rope_theta, DLTensor* o) {
+  // `tmp` is a 4096 * 4096 float16 scratch space.
```
Comment (yzh119) [inline src/tvm_wrapper.cu]:
```suggestion
  // `tmp` is a user provided scratch space.
```
`tmp` is used as a float32 tensor, and the required space is hardware dependent, 16mb is a suggested value.

File: src/tvm_wrapper.cu [hunk 2]
```diff
@@ -150,11 +127,10 @@ int _FlashInferSinglePrefillWithKVCache(DLTensor* q, DLTensor* k, DLTensor* v, b
...
 
-TVM_DLL_EXPORT_TYPED_FUNC(FlashInferSinglePrefillWithKVCache, _FlashInferSinglePrefillWithKVCache);
-
-int _FlashInferSingleDecodeWithKVCache(DLTensor* q, DLTensor* k, DLTensor* v, int64_t qkv_layout,
-                                       int64_t rotary_mode, double rope_scale, double rope_theta,
-                                       DLTensor* o) {
+int _FlashInferSingleDecodeWithKVCache(DLTensor* q, DLTensor* k, DLTensor* v, DLTensor* tmp,
+                                       int64_t qkv_layout, int64_t rotary_mode, double rope_scale,
+                                       double rope_theta, DLTensor* o) {
+  // `tmp` is a 4096 * 4096 float16 scratch space.
```
Comment (yzh119) [inline src/tvm_wrapper.cu]: ditto

**Comment by junrushao** [comment]
> Just want to confirm, `TVM_DLL_EXPORT_TYPED_FUNC` is no longer required, is that correct?

Yep correct! They are useful only when we are integrating them as a standalone .o file, rather than linked into TVM’s runtime

File: src/tvm_wrapper.cu (hunk 1 above)
Comment (junrushao) [reply src/tvm_wrapper.cu:82]: I see. Will use fp32 instead of fp16

Comment (junrushao) [reply src/tvm_wrapper.cu]: updated

**Review by yzh119 — APPROVED** [review]

**Pull request merged**